pkg/operator/ceph/cluster/osd/osd.go

```diff
@@ -193,16 +195,79 @@ func (c *Cluster) Start() error {
 	}
 	logger.Infof("wait timeout for healthy OSDs during upgrade or restart is %q", c.clusterInfo.OsdUpgradeTimeout)
 
+	osdsToSkipReconcile, err := c.getOSDsToSkipReconcile()
+	if err != nil {
+		logger.Warningf("failed to get osds to skip reconcile. %v", err)
+	}
+
 	// prepare for updating existing OSDs
 	updateQueue, deployments, err := c.getOSDUpdateInfo(errs)
 	if err != nil {
 		return errors.Wrapf(err, "failed to get information about currently-running OSD Deployments in namespace %q", namespace)
 	}
-	osdsToSkipReconcile, err := c.getOSDsToSkipReconcile()
-	if err != nil {
-		logger.Warningf("failed to get osds to skip reconcile. %v", err)
+
+	var replaceConfig *OSDReplaceConfig
+	if c.spec.Storage.Store.UpdateStore == OSDStoreUpdateConfirmation {
+		logger.Infof("start migrating osds to backend store %q", c.spec.Storage.Store.Type)
+
+		replaceConfig, err = c.getOSDReplaceConfig()
+		if err != nil {
+			return errors.Wrapf(err, "failed to get OSD replace config")
+		}
+
+		// skip updating the OSDs that need migration. This is to prevent the reconciler from updating the backend store in the OSD deployments
+		// that have not been migrated to the new store yet.
+		for _, osd := range replaceConfig.OSDs {
+			updateQueue.Remove([]int{osd.ID})
+		}
+
+		if len(replaceConfig.OSDs) != 0 {
+			logger.Infof("total osds that need migration to new backend store %q : %d", c.spec.Storage.Store.Type, len(replaceConfig.OSDs))
+			pgHealthMsg, pgClean, err := cephclient.IsClusterClean(c.context, c.clusterInfo)
+			if err != nil {
+				return errors.Wrapf(err, "failed to check if the cluster is clean")
+			}
+			logger.Infof("placement group status : %q", pgHealthMsg)
+
+			replaceConfig.IsPgHealthy = pgClean
+			if replaceConfig.IsPgHealthy {
+				if replaceConfig.OSDInProgress != nil {
+					// If OSDInProgress is still available in the config, then last reconcile might have failed to delete this OSD.
+					// So continue with the same OSD.
+					c.replaceOSD = replaceConfig.OSDInProgress
+				} else {
+					osd := replaceConfig.OSDs[0]
+					replaceConfig.OSDs = replaceConfig.OSDs[1:]
+
+					// TODO: should we check if the osd is ok-to-destroy before proceeding?
```

Comment:
Let's not call ok-to-destroy. We expect the OSD has PGs that will need to be anyway backfilled after the replacement. The critical thing is that we are active+clean before doing it.